### PR TITLE
Add system.rate-limit to configurable options

### PR DESCRIPTION
**Pull request by dcramer**
This includes various cleanup to forms, some fixes for the API endpoints to make them usable, and adds the system.rate-limit configuration option (to replace SENTRY_SYSTEM_MAX_EVENTS_PER_MINUTE).

### CODE REVIEW

**Comment by dcramer** [comment]
We need to make a decision on if options should allow NULL values. Right now this patch changes it so they can.

File: src/sentry/runner/initializer.py [hunk 1]
```diff
@@ -67,6 +67,7 @@ def get_asset_version(settings):
     'system.databases': 'DATABASES',
     'system.debug': 'DEBUG',
     'system.secret-key': 'SECRET_KEY',
+    'system.rate-limit': 'SENTRY_SYSTEM_MAX_EVENTS_PER_MINUTE',
```
Comment (mattrobenolt) [inline src/sentry/runner/initializer.py]: This shouldn't be done here.

Ideally, you'd just change every use of `SENTRY_SYSTEM_MAX_EVENTS_PER_MINUTE` instead to use the new options.

Then add an entry into `apply_legacy_settings` to move existing settings.SENTRY_SYSTEM_MAX_EVENTS_PER_MINUTE` into `SENTRY_OPTIONS`.

If we do it like this, the value won't change at runtime since it gets memoized into the settings value.

Comment (mattrobenolt) [inline src/sentry/runner/initializer.py]: To be clear, this mapping is here is purely for things that Django would rely on and some other edge cases in backwards compatibility that don't support an easy change to using options values.

**Comment by dcramer** [comment]
Alright fixed this up. I think things are good (I tested both new setup flow as well as settings page).

We backpedaled on nullable options, so 0 means disabled for rate limit (as it always did).

File: docs/throttling.rst [hunk 2]
```diff
@@ -50,12 +50,11 @@ configuration might look like this:
     }
 
 
-You can also configure system-wide maximums, and a default value for all projects:
+You can also configure the system-wide maximum per-minute rate limit:
 
 .. code-block:: python
 
-   SENTRY_DEFAULT_MAX_EVENTS_PER_MINUTE = '90%'
```
Comment (dcramer) [inline docs/throttling.rst]: this setting isnt really useful anymore, so we'll plan to kill it off

File: src/sentry/static/sentry/app/views/app.jsx [hunk 3]
```diff
@@ -73,7 +73,9 @@ const App = React.createClass({
       return (
         <div>
           <Indicators className="indicators-container" />
-          <InstallWizard onConfigured={this.onConfigured} />
+          <InstallWizard
+            onConfigured={this.onConfigured}
+            requiredOptions={ConfigStore.get('requiredOptions')} />
```
Comment (dcramer) [inline src/sentry/static/sentry/app/views/app.jsx]: oh need to kill this

File: src/sentry/conf/server.py [hunk 4]
```diff
@@ -740,9 +740,6 @@ def create_partitioned_queues(name):
 # The default value for project-level quotas
 SENTRY_DEFAULT_MAX_EVENTS_PER_MINUTE = '90%'
 
-# The maximum number of events per minute the system should accept.
-SENTRY_SYSTEM_MAX_EVENTS_PER_MINUTE = 0
```
Comment (mattrobenolt) [inline src/sentry/conf/server.py]: Do you want to add this into the `config.yml` now? Or just not surface this at all anymore?

File: src/sentry/options/manager.py [hunk 5]
```diff
@@ -178,8 +197,10 @@ def register(self, key, default='', type=None, flags=DEFAULT_FLAGS,
         # value instead that matches the type expected, rather than relying on None.
         if type is NoneType:
             raise TypeError('Options must not be NoneType')
-        if not isinstance(default, type):
+        if default is not None and not self._value_is_of_type(type, default):
             raise TypeError('got %r, expected %r' % (_type(default), type))
+        if default is None:
+            default = type()
```
Comment (mattrobenolt) [inline src/sentry/options/manager.py]: I guess this is a bit nicer.

Since we're doing this now, you can remove the `default={}` on the existing options.

File: src/sentry/options/manager.py [hunk 6]
```diff
@@ -81,7 +81,15 @@ def set(self, key, value):
         assert not (opt.flags & FLAG_PRIORITIZE_DISK and settings.SENTRY_OPTIONS.get(key)), '%r cannot be changed at runtime because it is configured on disk' % key
 
         if not isinstance(value, opt.type):
```
Comment (mattrobenolt) [inline src/sentry/options/manager.py]: You should use your `_value_is_of_type` here, no?

If we set it to basestring, you're saying the type becomes `unicode`, so what happens if it's being set as `str`? I don't think we'd want to lock into unicode vs str explicitly imo.

File: src/sentry/conf/server.py (hunk 4 above)
Comment (dcramer) [inline src/sentry/conf/server.py]: we dont need to add everything to config.yml imo

Comment (mattrobenolt) [inline src/sentry/conf/server.py]: Yeah, just asking. That's fine. :)

File: src/sentry/options/manager.py (hunk 6 above)
Comment (dcramer) [inline src/sentry/options/manager.py]: theres no reason that I can come up with to have multiple string types in options which is why its doing this

File: src/sentry/options/manager.py (hunk 5 above)
Comment (dcramer) [inline src/sentry/options/manager.py]: change is out of scope -- can always fix it when you change to immutable defaults

File: src/sentry/options/manager.py (hunk 6 above)
Comment (mattrobenolt) [inline src/sentry/options/manager.py]: Yeah, there's just lots of places where it's kinda unclear I guess whether a value is a string vs unicode. And this could trip us up here.

File: src/sentry/options/manager.py [hunk 7]
```diff
@@ -164,13 +172,24 @@ def delete(self, key):
...
                  ttl=DEFAULT_KEY_TTL, grace=DEFAULT_KEY_GRACE):
         assert key not in self.registry, 'Option already registered: %r' % key
         # Guess type based on the default value
         if type is None:
-            if isinstance(default, basestring):
-                type = basestring
+            # the default value would be equivilent to '' if no type / default
+            # is specified and we assume unicode for safety
+            if default is None:
+                type = unicode
```
Comment (mattrobenolt) [inline src/sentry/options/manager.py]: Shouldn't this just be left to basestring like it was? For the same reason as mentioned before. We can get tripped up with unicode vs str when we don't intend to.

Comment (dcramer) [inline src/sentry/options/manager.py]: you cannot basestring()

Comment (mattrobenolt) [inline src/sentry/options/manager.py]: I know, I was suggesting a similar dance like you do with `_value_is_of_type`.

But I can go through and fix this with better types after you're done.

I'm afraid of this breaking current implementation though if it's being typed as unicode, but we attempt to set it as a str, it'll be problematic.

**Comment by mattrobenolt** [comment]
Python looks good. :+1:
